### PR TITLE
add onCropsGrow event to KelpTopBlock

### PR DESCRIPTION
**Pull request by Bricktricker**
This adds the onCropsGrowPre and onCropsGrowPost event calls to the KelpTopBlock

### CODE REVIEW

File: patches/minecraft/net/minecraft/block/KelpTopBlock.java.patch [hunk 1]
```diff
@@ -0,0 +1,15 @@
+--- a/net/minecraft/block/KelpTopBlock.java
++++ b/net/minecraft/block/KelpTopBlock.java
+@@ -57,9 +57,12 @@
+       } else {
+          BlockPos blockpos = p_196267_3_.func_177984_a();
+          BlockState blockstate = p_196267_2_.func_180495_p(blockpos);
++         if(net.minecraftforge.common.ForgeHooks.onCropsGrowPre(p_196267_2_, blockpos, p_196267_1_, true)) {
```
Comment (Draco18s) [inline patches/minecraft/net/minecraft/block/KelpTopBlock.java.patch]: This is not how the crop growth events are supposed to be implemented. You are supposed to replace the existing if-check where the random value is, passing the result of the random check to the event, eg. line 60 should look like this:

    if (blockstate.getBlock() == Blocks.WATER && state.get(AGE) < 25 &&
        ForgeHooks.onCropsGrowPre(p_196267_2_, blockpos, p_196267_1_, random.nextDouble() < 0.14D))
    { ... }

The whole point is to be able to *force* the growth even if the vanilla logic says that the block does not grow *and* prevent it when vanilla logic says it does. Currently your patch only covers the latter.

**Review by Draco18s — CHANGES_REQUESTED** [review]

**Comment by Draco18s** [comment]
By the way, thanks for doing this. I had missed the kelp block when I was patching the sweet berry bush. I had tried to find any other new blocks that had an "age" property or implemented any of the other classes/interfaces that relate to growing blocks and managed to overlook this one.

File: patches/minecraft/net/minecraft/block/KelpTopBlock.java.patch (hunk 1 above)
Comment (Bricktricker) [reply patches/minecraft/net/minecraft/block/KelpTopBlock.java.patch]: Ok, I have updated the PR

File: patches/minecraft/net/minecraft/block/KelpTopBlock.java.patch [hunk 2]
```diff
@@ -1,15 +1,14 @@
...
        } else {
           BlockPos blockpos = p_196267_3_.func_177984_a();
           BlockState blockstate = p_196267_2_.func_180495_p(blockpos);
-+         if(net.minecraftforge.common.ForgeHooks.onCropsGrowPre(p_196267_2_, blockpos, p_196267_1_, true)) {
-          if (blockstate.func_177230_c() == Blocks.field_150355_j && p_196267_1_.func_177229_b(field_203163_a) < 25 && p_196267_4_.nextDouble() < 0.14D) {
+-         if (blockstate.func_177230_c() == Blocks.field_150355_j && p_196267_1_.func_177229_b(field_203163_a) < 25 && p_196267_4_.nextDouble() < 0.14D) {
++         if (blockstate.func_177230_c() == Blocks.field_150355_j && p_196267_1_.func_177229_b(field_203163_a) < 25 && net.minecraftforge.common.ForgeHooks.onCropsGrowPre(p_196267_2_, blockpos, p_196267_1_, p_196267_4_.nextDouble() < 0.14D)) {
              p_196267_2_.func_175656_a(blockpos, p_196267_1_.func_177231_a(field_203163_a));
           }
 +         net.minecraftforge.common.ForgeHooks.onCropsGrowPost(p_196267_2_, p_196267_3_, p_196267_1_);
```
Comment (Draco18s) [inline patches/minecraft/net/minecraft/block/KelpTopBlock.java.patch]: `onCropsGrowPost` should be up one line as well. It is only supposed to be fired on a successful growth.

**Review by Draco18s — CHANGES_REQUESTED** [review]

File: patches/minecraft/net/minecraft/block/KelpTopBlock.java.patch (hunk 2 above)
Comment (Bricktricker) [reply patches/minecraft/net/minecraft/block/KelpTopBlock.java.patch]: Missed that, moved it up into the if statement

**Review by Draco18s — APPROVED** [review]
Looks good now.

**Pull request merged**